### PR TITLE
KAFKA-5689:Add MeteredWindowStore and refactor store hierarchy

### PR DESCRIPTION
**Pull request by dguy**
Add MeteredWindowStore and ChangeLoggingWindowBytesStore.
Refactor Store hierarchy such that Metered is always the outermost store
Do serialization in MeteredWindowStore

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @enothereska @bbejeck

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6873/
Test PASSed (JDK 7 and Scala 2.11).

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java [hunk 1]
```diff
@@ -131,72 +136,69 @@ public void close() {
...
                                                                              timeTo);
         final PeekingKeyValueIterator<Bytes, LRUCacheEntry> filteredCacheIterator = new FilteredCacheIterator(
             cacheIterator, hasNextCondition, cacheFunction
         );
 
-        return new MergedSortedCacheWindowStoreIterator<>(filteredCacheIterator,
-                                                          underlyingIterator,
-                                                          new StateSerdes<>(serdes.topic(), Serdes.Long(), serdes.valueSerde()));
+        return new MergedSortedCacheWindowStoreIterator(filteredCacheIterator,
+                                                          underlyingIterator
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java]: nit: do we need multi-lines here?

**Review by guozhangwang — COMMENTED** [review]
LGTM! Could you make a double check with jconsole on the metric names and check if the state store metrics are still correct as stated in `ops.html`?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6859/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by dguy** [comment]
the metric names are the same as previous

**Comment by dguy** [comment]
merged to trunk

**Pull request closed without merging**